### ISSUE TITLE
Rationale for invoking continuations synchronously vs. asynchronously as part of a call to CancellationTokenSource.Cancel

### ISSUE THREAD

**Issue by WizardBrony** [dotnet/runtime#66354]
There was a recent change made to `Task.DelayPromiseWithCancellation` to invoke any continuations from the task as part of a `Cancel` call asynchronously:

https://github.com/dotnet/runtime/blob/446a0d610d4479fa9ba58ec14a89132a18ed33c4/src/libraries/System.Private.CoreLib/src/System/Threading/Tasks/Task.cs#L5639-L5662

However, there are other APIs that invoke continuations within a `Cancel` call synchronously, including the recently added `Task.CancellationPromise<TResult>` (for use by `Task[<TResult>].WaitAsync`):

https://github.com/dotnet/runtime/blob/446a0d610d4479fa9ba58ec14a89132a18ed33c4/src/libraries/System.Private.CoreLib/src/System/Threading/Tasks/Task.cs#L2823-L2830

I don't understand the reason for this discrepancy. Was this just missed, or is there a rationale for APIs like `Task[<TResult>].WaitAsync` to keep invoking continuations in `Cancel` calls synchronously?

**Comment by joperezr** [comment]
cc: @stephentoub

**Comment by tannergooding** [comment]
I don't believe there's any concrete reason for it differing here, it just didn't get the same treatment as #64217 (which was a targeted fix for #64051)

`Cancel` should be running registered callbacks synchronously, so any task so any task completed from within a cancellation callback will, by default, invoke its continuations synchronously on the thread calling `Cancel`. The singling out for `Task.Delay` and missing `WaitAsync` is likely just due to prevalence and how much it was hit by people.

I'm not convinced we need to make a similar change to `WaitAsync` as its default/de-facto behavior for everything else. If we had concrete reports, we might reconsider.

If you don't want arbitrary continuations running inline under `Cancel`, then don't call `Cancel` while holding a lock or ensure its called from a thread pool thread instead.
